### ISSUE TITLE
Traefik ACME DNS-01 gcloud duplicate records.

### ISSUE THREAD

**Issue by shanna** [containous/traefik#4149]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I'm not sure if this is a traefik or upstream bug but changing my acme provider from `route53` to `gcloud` my configuration generates `409` duplicate resource errors from google when generating certificates for two subdomains. Credentials are fine in both cases, the error appears isolated to the acme gcloud provider.

### What did you expect to see?

Successful certificate generation with `gcloud` provider. The configuration works with `route53` set as the provider.

### What did you see instead?

Error 409. It looks like the acme code expects to be able to upsert TXT records but googles API doesn't allow it. See log output below.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.4
Codename:     maroilles
Go version:   go1.11.1
Built:        2018-10-30_10:44:30AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
[acme]
  email = "email@mydomain.com"
  storage = "/acme-certificates/acme.json"
  acmeLogging = true
  entryPoint = "https"

[acme.dnsChallenge]
  provider = "gcloud"

[[acme.domains]]
  main = "*.development.mydomain.com"
  sans = ["development.mydomain.com"]

[[acme.domains]]
  main = "*.qa.mydomain.com"
  sans = ["qa.mydomain.com"]
```

### If applicable, please paste the log output in DEBUG level (`--logLevel=DEBUG` switch)

```
time="2018-11-01T11:46:41Z" level=error msg="Error obtaining certificate: acme: Error -> One or more domains had a problem:\n[qa.mydomain.com] error presenting token: googlecloud: googleapi: Error 409: The resource 'entity.change.additions[0].rrdata[1]' named '\"VuOLZS3tfLfrOswFtV3mayqOYaz5PvQK16nJdgMdMez\"' already exists, alreadyExists\n"
time="2018-11-01T11:46:41Z" level=error msg="Unable to obtain ACME certificate for domains \"*.qa.mydomain.com,qa.mydomain.com\" : unable to generate a certificate for the domains [*.qa.mydomain.com qa.mydomain.com]: acme: Error -> One or more domains had a problem:\n[qa.mydomain.com] error presenting token: googlecloud: googleapi: Error 409: The resource 'entity.change.additions[0].rrdata[1]' named '\"VuOLZS3tfLfrOswFtV3mayqOYaz5PvQK16nJdgMdMez\"' already exists, alreadyExists\n"
time="2018-11-01T11:46:41Z" level=error msg="Error obtaining certificate: acme: Error -> One or more domains had a problem:\n[development.mydomain.com] error presenting token: googlecloud: googleapi: Error 409: The resource 'entity.change.additions[0].rrdata[1]' named '\"Jfv6orVfbwi0IdQbmpvG3qnIltMX5x56vGUpHjoBzOa\"' already exists, alreadyExists\n"
time="2018-11-01T11:46:41Z" level=error msg="Unable to obtain ACME certificate for domains \"*.development.mydomain.com,development.mydomain.com\" : unable to generate a certificate for the domains [*.development.mydomain.com development.mydomain.com]: acme: Error -> One or more domains had a problem:\n[development.mydomain.com] error presenting token: googlecloud: googleapi: Error 409: The resource 'entity.change.additions[0].rrdata[1]' named '\"Jfv6orVfbwi0IdQbmpvG3qnIltMX5x56vGUpHjoBzOa\"' already exists, alreadyExists\n"
```

**Comment by MrFlutters** [comment]
i have had this same kind of issue recently on gcloud, my current workaround is adding `--acme.dnschallenge.delaybeforecheck=180"` or if you use traefik.toml
```
[acme.dnsChallenge]
delayBeforeCheck = 180
```

it seems to try and set the TXT record multiple times, By setting a 3 minute delay, it seems to fix this.


By default the TXT record's TTL on google is set to 2 minutes, by delaying the dns check it seems to solve this issue.

My guess would be it's attempting to do the dns certificate mutliple times at startup but TTL caching the incorrect token but having the correct token in DNS.

**Comment by danacr** [comment]
I experienced the same issue, but I only managed to get it working by setting the delay to 600 seconds.

**Comment by shanna** [comment]
I imagine since the provider already has the ability to read DNS records it could look to see if there is an existing TXT record and use the TTL to set an appropriate default delay rather than making this a manual process or magic number. I haven't looked at the provider code or the limitations of googles DNS API though, I don't have time to contribute more than a bug report at the moment.

**Comment by MrFlutters** [comment]
@shanna the gcloud provided never used to require a delay, it seems something changed recently on google's end, as the traefik version i used previously also has this bug.

@danacr you shouldn't need to set the delay that high, unless you kept restarting traefik or having issues, usually waiting a couple of minutes **and** making sure a record isnt still in cloud dns lets u go back to 180 seconds.

**Comment by tompson** [comment]
but as @shanna suggested, the best solution would be to adapt the provider implementation so that it checks if the required TXT record already exists ... maybe that is something I could implement an create a PR

**Comment by MrFlutters** [comment]
@tompson 
if im right, traefik uses xenolf/lego [here](https://github.com/containous/traefik/blob/c39d21c1783d1d24b100448cffc00397f6a5fe63/provider/acme/provider.go#L28) for most of its providers [here](https://github.com/xenolf/lego/tree/master/providers/dns/)

testing gcloud on lego, by default adds a timeout for 3 minutes (added by [this commit](https://github.com/xenolf/lego/pull/634/commits/a707ae78eb176a138a86be81a3ffad3c4a07700e), and does not suffer the same issue,

its possible traefik isn't taking into account the default propagation timer and/or setting it to 0
```
2018/11/27 14:27:50 [INFO] [*.example.com, example.com] acme: Obtaining bundled SAN certificate
2018/11/27 14:27:51 [INFO] [*.example.com] AuthURL: [REDACTED]
2018/11/27 14:27:51 [INFO] [example.com] AuthURL: [REDACTED]
2018/11/27 14:27:51 [INFO] [example.com] acme: Could not find solver for: http-01
2018/11/27 14:27:51 [INFO] [example.com] acme: Preparing to solve DNS-01
2018/11/27 14:27:55 [INFO] [example.com] acme: Preparing to solve DNS-01
2018/11/27 14:27:58 [INFO] [example.com] acme: Trying to solve DNS-01
2018/11/27 14:27:58 [INFO] [example.com] Checking DNS record propagation using [127.0.0.53:53]
2018/11/27 14:27:58 [INFO] Wait [timeout: 3m0s, interval: 5s]
2018/11/27 14:28:44 [INFO] [example.com] The server validated our request
2018/11/27 14:28:44 [INFO] [example.com] acme: Trying to solve DNS-01
2018/11/27 14:28:44 [INFO] [example.com] Checking DNS record propagation using [127.0.0.53:53]
2018/11/27 14:28:44 [INFO] Wait [timeout: 3m0s, interval: 5s]
2018/11/27 14:29:11 [INFO] [example.com] The server validated our request
2018/11/27 14:29:14 [INFO] [*.example.com, example.com] acme: Validations succeeded; requesting certificates
2018/11/27 14:29:15 [INFO] [*.example.com] Server responded with a certificate.
```

**Comment by danacr** [comment]
In our use case, it happens quite often to restart traefik and it breaks when detecting existing records on GCP.

**Comment by MrFlutters** [comment]
i think the issue is with the dns challenge token check rather than an issue with the change sets. 
seems to check if the current token already exists, but it cant seems to be able to use that token and when a new token is generated, google dns wont update in time due to caching, causing a loop

**Comment by danacr** [comment]
`level=error msg="Unable to obtain ACME certificate for domains \"domain.example.org\" : unable to generate a certificate for the domains [domain.example.org]: acme: Error -> One or more domains had a problem:\n[domain.example.org] error presenting token: googlecloud: googleapi: Error 409: The resource 'entity.change.additions[0]' named '_acme-challenge.domain.example.org. (TXT)' already exists, alreadyExists\n"`

It seems to be the same error both on restart of traefik and when there is no delay in dns check.

**Comment by MrFlutters** [comment]
@ldez it seems to be working fine on my end, surviving multiple container restarts without the delay and no longer any issues for me with gcloud

**Comment by ldez** [comment]
@MrFlutters I updated the image, could you pull and retry?

**Comment by MrFlutters** [comment]
@ldez im still seeing the issue, and seems GCE_PROPAGATION_TIMEOUT does not resolve it

**Comment by ldez** [comment]
`delayBeforeCheck` and  `GCE_PROPAGATION_TIMEOUT` are related to the issue.

Have you use my new image?

**Comment by ldez** [comment]
Could you give the error log.

**Comment by MrFlutters** [comment]
sorry, must've messed something up, its working now, certificates generated successfully, however some errors still get logged

`Error 404: The 'entity.change.deletions[0]' resource named '_acme-challenge.sub2.example.com. (TXT)' does not exist., notFound\n"`

full log provided if curious.
[traefik.log](https://github.com/containous/traefik/files/2666563/traefik.log)

**Comment by ldez** [comment]
I tried a new approach and updated the image, could you pull and retry?

**Comment by MrFlutters** [comment]
Ok, Testing the new image still shows gcloud errors and it reports recieving a certificate. however traefik now responds with the default traefik cert for both wildcards, rather than the generated certs.
(acme.json is populated with generated certs)

`level=debug msg="Serving default cert for request: \"nginx.sub1.example.com\""`
`level=debug msg="Serving default cert for request: \"nginx.sub2.example.com\""`

___
[traefik3.log](https://github.com/containous/traefik/files/2670717/traefik3.log)

**Comment by MrFlutters** [comment]
Testing the new image, `*.sub1.example.com` generates a usable certificate, However `*.sub2.example.com` still uses the default.

Both domain report as `Server responded with a certificate`

I am using a simple nginx container(with 3 replicas each) to test and i have also kept the container logs in this file.
___
[traefik4.log](https://github.com/containous/traefik/files/2670921/traefik4.log)

**Comment by MrFlutters** [comment]
@ldez ah,I didnt notice that. I'm currently using the command line for configuration,
i've used
`"--acme.domains=*.sub1.example.com,sub1.example.com"`
`"--acme.domains=*.sub2.example.com,sub2.example.com"`
from the built in --help flag
___
However, sub1 wildcards work, however sub2 wildcards still returns the default cert
___
[traefik5.log](https://github.com/containous/traefik/files/2671044/traefik5.log)
___
this is my current configuration for clarity.

`ldez/traefik:gcloud \`
`-l DEBUG \`
`--docker.exposedbydefault=false \`
`--docker.domain=sub.example.com \`
`--docker.network=external \`
`--docker.watch \`
`--docker.swarmmode \`
`--api  \`
`"--entryPoints=Name:http Address::80" \`
`"--entryPoints=Name:https Address::443 TLS" \`
`"--defaultentrypoints=http,https" \`
`"--acme" \`
`"--acme.acmelogging" \`
`"--acme.entrypoint=https" \`
`"--acme.storage=/etc/traefik/acme.json" \`
`"--acme.dnschallenge" \`
`"--acme.dnschallenge.provider=gcloud" \`
`"--acme.dnschallenge.resolvers=8.8.8.8" \`
`"--acme.domains=*.sub1.example.com,sub1.example.com" \`
`"--acme.domains=*.sub2.example.com,sub2.example.com" \`
`"--acme.email=example@gmail.com" \`
`"--acme.onhostrule"`

**Comment by ldez** [comment]
could you try with `--acme.onhostrule=false`?

**Comment by MrFlutters** [comment]
setting `onhostrule=false` seems to have stopped the certificates being able to generate,
I get `level=error msg="Error obtaining certificate retrying in 364.610588ms"` for both subdomains
___
[traefik6.log](https://github.com/containous/traefik/files/2671194/traefik6.log)

**Comment by MrFlutters** [comment]
@ldez i just tried the configuration again after leaving it for awhile and recreating the container.
it seems to be working now.

**Comment by ldez** [comment]
Traefik don't reload. I think you have another issue.
So I propose to continue the discussion on Slack.

For others people, could you try my latest image and give me some returns.

**Comment by danacr** [comment]
@ldez I keep getting the error:

`time="2018-12-18T05:08:43Z" level=error msg="Unable to obtain ACME certificate for domains \"*.test.DOMAIN.io\" : unable to generate a certificate for the domains [*.test.DOMAIN.io]: acme: Error -> One or more domains had a problem:\n[test.DOMAIN.io] error presenting token: googlecloud: failed to perform changes [zone DOMAIN, change {\"additions\":[{\"name\":\"_acme-challenge.test.DOMAIN.io.\",\"rrdatas\":[\"-3xolfqSDzC6ZSBTobMFcG6w4iox_F363UuMm3Ki4E0\"],\"ttl\":120,\"type\":\"TXT\"}]}]: googleapi: Error 409: The resource 'entity.change.additions[0]' named '_acme-challenge.test.DOMAIN.io. (TXT)' already exists, alreadyExists\n"`

**Comment by ldez** [comment]
@danacr could you give your configuration (TOML, CLI, docker-compose, ...) and the full logs?

**Comment by danacr** [comment]
command:
       - "--entryPoints=Name:http Address::80 Redirect.EntryPoint:https"
       - "--entrypoints=Name:https Address::443 TLS"
       - --api
       - docker
       - --docker.endpoint=unix:///var/run/docker.sock
       - --acme=true
       - --acme.dnschallenge
       - --acme.dnschallenge.provider=gcloud
      # - --acme.dnschallenge.delaybeforecheck=300
       - --acme.domains=$DOMAIN
       - --acme.domains=*.$DOMAIN
       - --acme.email=dan@kryha.io
       - --acme.entrypoint=https
       - --acme.storage=/etc/traefik-certs/acme.json
    labels:
       - traefik.backend=traefik
       - traefik.port=8080
       - traefik.frontend.auth.basic=user:stuff
       - traefik.enable=true
       - traefik.frontend.entryPoints=https
       - traefik.frontend.rule=Host:traefik.$DOMAIN
    ports:
      - 80:80
      - 443:443
    networks:
      - traefik
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - ./certs:/etc/traefik-certs


Update: I recreated it and now it works

`traefik           | time="2018-12-18T09:23:36Z" level=error msg="Error getting levelnot a valid logrus Level: \"\\\"debug\\\"\""`

Apparently they way I specified  - --logLevel="DEBUG" did not work

**Comment by ldez** [comment]
could you change your configuration to this:
```
- --acme.domains=*.$DOMAIN,$DOMAIN
``

**Comment by ldez** [comment]
use `"--acme.domains=*.sub2.example.com,sub2.example.com" `

**Comment by danacr** [comment]
traefik:
    image: ldez/traefik:gcloud
    command:
       - "--entryPoints=Name:http Address::80 Redirect.EntryPoint:https"
       - "--entrypoints=Name:https Address::443 TLS"
       - --api
       - docker
       - --docker.endpoint=unix:///var/run/docker.sock
       - --acme=true
       - --acme.dnschallenge
       - --acme.dnschallenge.provider=gcloud
       - "--acme.domains=*.$DOMAIN,$DOMAIN"
       - --acme.email=dan@kryha.io
       - --acme.entrypoint=https
       - --acme.storage=/etc/traefik-certs/acme.json

Confirmed working on an instance. Will try again on a clean one to confirm.

**Comment by danacr** [comment]
@ldez I can confirm ldez/traefik:gcloud works as expected using the above-mentioned command line arguments

**Comment by ldez** [comment]
@danacr could you retry with the standard docker image?

**Comment by danacr** [comment]
I can confirm that it works with the following configuration:

    image: traefik:latest
    command:
      - "--entryPoints=Name:http Address::80 Redirect.EntryPoint:https"
      - "--entrypoints=Name:https Address::443 TLS"
      - --api
      - --docker
      - --docker.endpoint=unix:///var/run/docker.sock
      - --acme
      - --acme.dnschallenge
      - --acme.dnschallenge.provider=gcloud
      - "--acme.domains=*.$DOMAIN,$DOMAIN"
      - --acme.email=dan@kryha.io
      - --acme.entrypoint=https
      - --acme.storage=/etc/traefik-certs/acme.json

**Comment by ldez** [comment]
I finally found the root of the bug, a PR is coming really soon.

**Comment by traefiker** [comment]
Closed by #4277.
